### PR TITLE
Improved the ray tracing fallback mechanic (case 1350590).

### PR DESCRIPTION
**Pull request by anisunity**
https://fogbugz.unity3d.com/f/cases/1350590/
Added a control for the fallback for the last bounce (rtgi/rtr/rr) to keep a previously existing side effect on user demand (case 1350590).
- Fixed an inconsistance between perf mode and quality mode for sky lighting (case 1350590).
- Fixed an inconsistance between perf quand quality mode for material simplification in rtgi (case 1350590).
This PR adds a lot of tests to significantly increase the coverage.
image
image
image

Testing status
Tests have been updated and are all green locally.

### CODE REVIEW

**Review by iM0ve — CHANGES_REQUESTED** [review]
Whats tested:
* Used the new parameters to see if they function as expected
* Upgraded 2 projects to see that the result with default values is unchanged (except bug fix)
* Verified that SSGI, SSR and Recursive rendering is still working properly

Minor concerns:
* Naming of the new properties does not match exactly on all overrides
image
* Tooltip is missing for Recursive Rendering

https://user-images.githubusercontent.com/36502659/128500397-317611df-96e5-4757-8233-8f792aae3de6.mp4

**Demo:**

https://user-images.githubusercontent.com/36502659/128500455-2f097060-f887-4543-9c66-06cacbeb0559.mp4

**Review by iM0ve — APPROVED** [review]
Verified the latest fixes and quickly tested properties still work

**Review by JulienIgnace-Unity — APPROVED** [review]

File: com.unity.render-pipelines.high-definition/Documentation~/Override-Screen-Space-GI.md [hunk 1]
```diff
@@ -46,13 +46,16 @@ HDRP uses the [Volume](Volumes.md) framework to calculate SSGI, so to enable and
...
 | **Depth Tolerance** | Use the slider to control the tolerance when comparing the depth of the GameObjects on screen and the depth buffer. Because the SSR algorithm can not distinguish thin GameObjects from thick ones, this property helps trace rays behind GameObjects. The algorithm applies this property to every GameObject uniformly. |
+| **Ray Miss**         | Defines if HDRP should use the reflection probes, the sky, both or nothing as a fall-back for screen space global illumination when a ray doesn't find an intersection. |
 
 ### Ray-traced
 
 ![](Images/Override-ScreenSpaceGlobalIllumination2.png)
 
 | Property                       | Description                                                  |
 | ------------------------------ | ------------------------------------------------------------ |
+| **Ray Miss**                   | Defines if HDRP should use the reflection probes, the sky, both or nothing as a fall-back for ray-traced global illumination when a ray doesn't find an intersection. |
```
Comment (Vic-Cooper) [inline com.unity.render-pipelines.high-definition/Documentation~/Override-Screen-Space-GI.md:57]: Determines what HDRP does when ray-traced global illumination (RTGI) doesn't find an intersection. . Choose from one of the following options: <br/>
• **Reflection probes**: HDRP uses reflection probes in your scene to calculate the last RTGI bounce.<br/>
• **Sky**: HDRP uses the sky defined by the current [Volume](Volumes.md) settings to calculate the last SSGI bounce.<br/>
• **Both** : HDRP uses both reflection probes and the  the sky defined by the current [Volume](Volumes.md) settings to calculate the last SSGI bounce.<br/>
• **Nothing**: HDRP does not calculate indirect lighting when RTGI doesn't find an intersection.<br/><br/>

This property is set to **Both** by default

File: com.unity.render-pipelines.high-definition/Documentation~/Override-Screen-Space-GI.md [hunk 2]
```diff
@@ -46,13 +46,16 @@ HDRP uses the [Volume](Volumes.md) framework to calculate SSGI, so to enable and
...
+| **Last Bounce**                | Defines if HDRP should use the reflection probes, the sky, both or nothing as a fall-back for ray-traced global illumination when lighting the last bounce. |
```
Comment (Vic-Cooper) [inline com.unity.render-pipelines.high-definition/Documentation~/Override-Screen-Space-GI.md:58]: Determines what HDRP does when ray-traced global illumination (RTGI) lights the last bounce. Choose from one of the following options: <br/>
• **Reflection probes**: HDRP uses reflection probes in your scene to calculate the last RTGI bounce.<br/>
• **Sky**: HDRP uses the sky defined by the current [Volume](Volumes.md) settings to calculate the last RTGI bounce.<br/>
• **Both** :  HDRP uses both reflection probes and the sky defined by the current [Volume](Volumes.md) settings to calculate the last RTGI bounce.<br/>
• **Nothing**: HDRP does not calculate indirect lighting when it evaluates the last bounce.<br/><br/>

This property is set to **Both** by default.

**Review by Vic-Cooper — CHANGES_REQUESTED** [review]
Docs reviewed, please check for accuracy.

**Comment by anisunity** [comment]
Just pushed the docs updates

File: com.unity.render-pipelines.high-definition/Documentation~/Override-Screen-Space-GI.md [hunk 3]
```diff
@@ -46,13 +46,16 @@ HDRP uses the [Volume](Volumes.md) framework to calculate SSGI, so to enable and
...
 
 ### Ray-traced
 
 ![](Images/Override-ScreenSpaceGlobalIllumination2.png)
 
 | Property                       | Description                                                  |
 | ------------------------------ | ------------------------------------------------------------ |
+| **Ray Miss**                   | Determines what HDRP does when ray-traced global illumination (RTGI) doesn't find an intersection. Choose from one of the following options: <br/>&#8226;**Reflection probes**: HDRP uses reflection probes in your scene to calculate the last RTGI bounce.<br/>&#8226;**Sky**: HDRP uses the sky defined by the current [Volume](Volumes.md) settings to calculate the last SSGI bounce.<br/>&#8226;**Both** : HDRP uses both reflection probes and the  the sky defined by the current [Volume](Volumes.md) settings to calculate the last SSGI bounce.<br/>&#8226;**Nothing**: HDRP does not calculate indirect lighting when RTGI doesn't find an intersection.<br/><br/>This property is set to **Both** by default|
```
Comment (Vic-Cooper) [inline com.unity.render-pipelines.high-definition/Documentation~/Override-Screen-Space-GI.md:57]: SSGI should be RTGI here, apologies for missing that!: 

 Determines what HDRP does when ray-traced global illumination (RTGI) doesn't find an intersection. Choose from one of the following options: <br/>&#8226;**Reflection probes**: HDRP uses reflection probes in your scene to calculate the last RTGI bounce.<br/>&#8226;**Sky**: HDRP uses the sky defined by the current [Volume](Volumes.md) settings to calculate the last RTGI bounce.<br/>&#8226;**Both** : HDRP uses both reflection probes and the  the sky defined by the current [Volume](Volumes.md) settings to calculate the last RTGI bounce.<br/>&#8226;**Nothing**: HDRP does not calculate indirect lighting when RTGI doesn't find an intersection.<br/><br/>This property is set to **Both** by default

**Review by Vic-Cooper — CHANGES_REQUESTED** [review]
SSGI > RTGI typos.

File: com.unity.render-pipelines.high-definition/Documentation~/Ray-Tracing-Recursive-Rendering.md [hunk 4]
```diff
@@ -29,7 +29,7 @@ You can also do this for Shader Graph master nodes:
 1. In the Project window, double-click on the Shader to open it in Shader Graph.
 2. In the **Graph Settings** tab of the **Graph Inspector**, go to **Surface Options** and enable **Recursive Rendering (Preview)**.
 
-It is best practice to use recursive rendering in situations that require multi-bounced reflection and refraction, for example car headlights. Although recursive rendering also produces ray-traced reflections in more simple scenarios, like for a mirror or a puddle, it is best practice to use [ray-traced reflections](Ray-Traced-Reflections.md) here for performance reasons.
+It is best practice to use recursive rendering in situations that require multi-bounced reflection and refraction, for example car headlights. Although recursive rendering also produces recursive rendering in more simple scenarios, like for a mirror or a puddle, it is best practice to use [ray-traced reflections](Ray-Traced-Reflections.md) here for performance reasons.
```
Comment (Vic-Cooper) [inline com.unity.render-pipelines.high-definition/Documentation~/Ray-Tracing-Recursive-Rendering.md:32]: Let's improve the phrasing here: 

It is best practice to use recursive rendering in situations that require multi-bounced reflection and refraction, for example, car headlights. You can use recursive rendering in simple scenarios, like a mirror or a puddle, but for best performance, use [ray-traced reflections](Ray-Traced-Reflections.md).

File: com.unity.render-pipelines.high-definition/Documentation~/Ray-Tracing-Recursive-Rendering.md [hunk 5]
```diff
@@ -41,3 +41,5 @@ Since recursive rendering uses an independent render pass, HDRP cannot render an
...
 | **Max Ray Length** | Controls the length of the rays that HDRP uses for ray tracing. If a ray doesn't find an intersection, then the ray returns the color of the sky. |
 | **Min Smoothness** | Defines the threshold at which reflection rays are not cast if the smoothness value of the target surface is inferior to the one defined by the parameter. |
+| **Ray Miss**                  | Determines what HDRP does when recursive rendering (RR) doesn't find an intersection. Choose from one of the following options: <br/>&#8226;**Reflection probes**: HDRP uses reflection probes in your scene to calculate the last RR bounce.<br/>&#8226;**Sky**: HDRP uses the sky defined by the current [Volume](Volumes.md) settings to calculate the last RR bounce.<br/>&#8226;**Both** : HDRP uses both reflection probes and the  the sky defined by the current [Volume](Volumes.md) settings to calculate the last RR bounce.<br/>&#8226;**Nothing**: HDRP does not calculate indirect lighting when RR doesn't find an intersection.<br/><br/>This property is set to **Both** by default|
```
Comment (Vic-Cooper) [inline com.unity.render-pipelines.high-definition/Documentation~/Ray-Tracing-Recursive-Rendering.md:44]: I don't think we need to abbreviate recursive rendering to RR here (also fixed double "the" typo): 

| **Ray Miss**                  | Determines what HDRP does when recursive rendering doesn't find an intersection. Choose from one of the following options: <br/>&#8226;**Reflection probes**: HDRP uses reflection probes in your scene to calculate the last recursive rendering bounce.<br/>&#8226;**Sky**: HDRP uses the sky defined by the current [Volume](Volumes.md) settings to calculate the last recursive rendering bounce.<br/>&#8226;**Both** : HDRP uses both reflection probes and the sky defined by the current [Volume](Volumes.md) settings to calculate the last recursive rendering bounce.<br/>&#8226;**Nothing**: HDRP does not calculate indirect lighting when recursive rendering doesn't find an intersection.<br/><br/>This property is set to **Both** by default|

File: com.unity.render-pipelines.high-definition/Documentation~/Ray-Tracing-Recursive-Rendering.md [hunk 6]
```diff
@@ -41,3 +41,5 @@ Since recursive rendering uses an independent render pass, HDRP cannot render an
...
+| **Last Bounce**               | Determines what HDRP does when recursive rendering (RR) lights the last bounce. Choose from one of the following options: <br/>&#8226;**Reflection probes**: HDRP uses reflection probes in your scene to calculate the last RR bounce.<br/>&#8226;**Sky**: HDRP uses the sky defined by the current [Volume](Volumes.md) settings to calculate the last RR bounce.<br/>&#8226;**Both**:  HDRP uses both reflection probes and the sky defined by the current [Volume](Volumes.md) settings to calculate the last RR bounce.<br/>&#8226;**Nothing**: HDRP does not calculate indirect lighting when it evaluates the last bounce.<br/><br/>This property is set to **Both** by default. |
```
Comment (Vic-Cooper) [inline com.unity.render-pipelines.high-definition/Documentation~/Ray-Tracing-Recursive-Rendering.md:45]: As above: 

| **Last Bounce**               | Determines what HDRP does when recursive rendering lights the last bounce. Choose from one of the following options: <br/>&#8226;**Reflection probes**: HDRP uses reflection probes in your scene to calculate the last bounce.<br/>&#8226;**Sky**: HDRP uses the sky defined by the current [Volume](Volumes.md) settings to calculate the last recursive rendering bounce.<br/>&#8226;**Both**:  HDRP uses both reflection probes and the sky defined by the current [Volume](Volumes.md) settings to calculate the last recursive rendering bounce.<br/>&#8226;**Nothing**: HDRP does not calculate indirect lighting when it evaluates the last bounce.<br/><br/>This property is set to **Both** by default. |

**Review by Vic-Cooper — CHANGES_REQUESTED** [review]
Small changes in ray-tracing-recursive-rendering.

File: com.unity.render-pipelines.high-definition/Documentation~/Ray-Tracing-Recursive-Rendering.md (hunk 4 above)
Comment (anisunity) [reply com.unity.render-pipelines.high-definition/Documentation~/Ray-Tracing-Recursive-Rendering.md:32]: not sure when i changed this, probably a find and replace

**Review by Vic-Cooper — APPROVED** [review]
Docs done, thanks!

**Pull request merged**